pkg/kratos/handlers.go

```diff
@@ -261,17 +286,66 @@ func (a *API) handleUpdateFlow(w http.ResponseWriter, r *http.Request) {
 	setCookies(w, cookies)
 
 	if shouldRegenerateBackupCodes {
-		a.lookupSecretsSettingsRedirect(w, flowId, *loginFlow.ReturnTo, loginFlow.GetOauth2LoginChallenge())
+		a.lookupSecretsSettingsRedirect(w, r, flowId, *loginFlow.ReturnTo, flowCookie)
 		return
 	}
 
 	if shouldEnforceMfa {
-		a.mfaSettingsRedirect(w, *loginFlow.ReturnTo, loginFlow.GetOauth2LoginChallenge())
+		a.mfaSettingsRedirect(w, r, *loginFlow.ReturnTo, flowCookie)
+		return
+	}
+
+	if redirectTo != nil {
+		a.cookieManager.SetStateCookie(w, flowCookie)
+		a.backchannelRedirectResponse(w, r, redirectTo)
+		return
+	}
+
+	// This is a hydra flow
+	if lc != "" {
+		// User is authenticated, return the session
+		response, cookies, err := a.handleCreateFlowWithSession(r, &flow.Session, lc)
+		if err != nil {
+			a.logger.Errorf(err.Error())
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		a.cookieManager.ClearStateCookie(w)
+		setCookies(w, cookies)
+		a.backchannelRedirectResponse(w, r, response)
+		return
+	}
+
+	resp, err := flow.MarshalJSON()
+	if err != nil {
+		a.logger.Errorf("Error when marshalling Json: %v\n", err)
+		http.Error(w, "Failed to parse flow error", http.StatusInternalServerError)
 		return
 	}
 
 	w.WriteHeader(http.StatusOK)
-	_ = json.NewEncoder(w).Encode(flow)
+	w.Write(resp)
+}
+
+// TODO: Refactor me for IAM-1286
+func (a *API) backchannelRedirectResponse(w http.ResponseWriter, r *http.Request, resp RedirectToInterface) {
```

Comment:
random q: why `backchannel`?

Reply:
Wasn't sure what name to use. I wanted this to be used to handle redirect responses for XHR requests (which are back-channel requests), BUT given the limitations of our UI this will have to handle browser redirects as well. I will rename it.